src/Symfony/Component/Semaphore/Store/RedisStore.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Semaphore\Store;
+
+use Symfony\Component\Cache\Traits\RedisClusterProxy;
+use Symfony\Component\Cache\Traits\RedisProxy;
+use Symfony\Component\Semaphore\Exception\InvalidArgumentException;
+use Symfony\Component\Semaphore\Exception\SemaphoreAcquiringException;
+use Symfony\Component\Semaphore\Exception\SemaphoreExpiredException;
+use Symfony\Component\Semaphore\Key;
+use Symfony\Component\Semaphore\PersistingStoreInterface;
+
+/**
+ * RedisStore is a PersistingStoreInterface implementation using Redis as store engine.
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RedisStore implements PersistingStoreInterface
+{
+    private $redis;
+
+    /**
+     * @param \Redis|\RedisArray|\RedisCluster|\Predis\ClientInterface $redisClient
+     */
+    public function __construct($redisClient)
+    {
+        if (!$redisClient instanceof \Redis && !$redisClient instanceof \RedisArray && !$redisClient instanceof \RedisCluster && !$redisClient instanceof \Predis\ClientInterface && !$redisClient instanceof RedisProxy) {
+            throw new InvalidArgumentException(sprintf('%s() expects parameter 1 to be Redis, RedisArray, RedisCluster or Predis\ClientInterface, %s given', __METHOD__, \is_object($redisClient) ? \get_class($redisClient) : \gettype($redisClient)));
+        }
+
+        $this->redis = $redisClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key, float $ttlInSecond)
+    {
+        $script = '
+            local key = KEYS[1]
+            local counterKey = key .. ":counter"
+            local ownerKey = key .. ":owner"
+            local timeKey = key .. ":time"
+            local identifier = ARGV[1]
+            local now = tonumber(ARGV[2])
+            local ttlInSecond = tonumber(ARGV[3])
+            local limit = tonumber(ARGV[4])
+
+            -- Remove expired values
+            redis.call("ZREMRANGEBYSCORE", timeKey, "-inf", now)
+            redis.call("ZINTERSTORE", ownerKey, 2, ownerKey, timeKey, "WEIGHTS", 1, 0)
+
+            -- Try to get a semaphore
+            local counter = redis.call("INCR", counterKey)
+            redis.call("ZADD", timeKey, now + ttlInSecond, identifier)
+            redis.call("ZADD", ownerKey, counter, identifier)
+            local rank = redis.call("ZRANK", ownerKey, identifier)
+
+            if rank < limit then
+                return true
+            end
+
+            redis.call("ZREM", timeKey, identifier)
+            redis.call("ZREM", ownerKey, identifier)
```

Comment:
Why not?

```suggestion
            local rank = redis.call("ZRANK", ownerKey, identifier)

            if rank + 1 < limit then
                redis.call("ZADD", timeKey, now + ttlInSecond, identifier)
                redis.call("ZADD", ownerKey, counter, identifier)
                return true
            end
```

Reply:
I implemented what is described in the book.  I guess the know what they are doing.

But at the first glance, I would say there is a possible race condition in you implementation.
But 1/ since we lock the transaction 2/ we use lua (so one transaction at a time), I think your code is valid.

I will clean other part of the code for know (+ add tests).  But I keep this comment open.

Thanks

Reply:
I added lots of test. And with your patches there are broken. Sorry
I'm going to close this feedback  - but feel free to submit a PR later